cloud/ec2_fleet.go

```diff
@@ -524,69 +537,17 @@ func (m *ec2FleetManager) requestFleet(ctx context.Context, ec2Settings *EC2Prov
 
 func (m *ec2FleetManager) makeOverrides(ctx context.Context, ec2Settings *EC2ProviderSettings) ([]*ec2.FleetLaunchTemplateOverridesRequest, error) {
 	subnets := m.settings.Providers.AWS.Subnets
-	if len(subnets) > 0 {
-		overrides := make([]*ec2.FleetLaunchTemplateOverridesRequest, 0, len(subnets))
-		for _, subnet := range subnets {
-			supported, err := typeCache.azSupportsInstanceType(ctx, m.client, subnet.AZ, ec2Settings.InstanceType)
-			if err != nil {
-				return nil, errors.Wrapf(err, "can't get supported instance types for AZ '%s'", subnet.AZ)
-			}
-			if supported {
-				overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{SubnetId: aws.String(subnet.SubnetID)})
-			}
-		}
-
-		return overrides, nil
+	if len(subnets) == 0 {
+		return nil, errors.New("no AWS subnets were configured")
 	}
 
-	describeVpcsOutput, err := m.client.DescribeVpcs(ctx, &ec2.DescribeVpcsInput{
-		Filters: []*ec2.Filter{
-			{
-				Name: aws.String("tag:Name"),
-				Values: []*string{
-					aws.String(ec2Settings.VpcName),
-				},
-			},
-		},
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "error finding vpc id")
-	}
-	err = validateEc2DescribeVpcsOutput(describeVpcsOutput)
-	if err != nil {
-		return nil, errors.Wrap(err, "invalid describe VPCs response")
-	}
-
-	describeSubnetsOutput, err := m.client.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
-		Filters: []*ec2.Filter{
-			{
-				Name:   aws.String("vpc-id"),
-				Values: []*string{describeVpcsOutput.Vpcs[0].VpcId},
-			},
-		},
-	})
-	if err != nil {
-		return nil, errors.Wrapf(err, "can't get subnets for vpc '%s'", ec2Settings.VpcName)
-	}
-	err = validateEc2DescribeSubnetsOutput(describeSubnetsOutput)
+	overrides := make([]*ec2.FleetLaunchTemplateOverridesRequest, 0, len(subnets))
+	supportingSubnets, err := typeCache.subnetsWithInstanceType(ctx, m.settings, m.client, ec2Settings.InstanceType)
 	if err != nil {
-		return nil, errors.Wrap(err, "invalid describe subnets response")
+		return nil, errors.Wrapf(err, "can't get AZs supporting instance type '%s'", ec2Settings.InstanceType)
 	}
-
-	AZSet := make(map[string]bool)
-	overrides := make([]*ec2.FleetLaunchTemplateOverridesRequest, 0, len(describeSubnetsOutput.Subnets))
-	for _, subnet := range describeSubnetsOutput.Subnets {
-		// AWS only allows one override per AZ
-		if !AZSet[*subnet.AvailabilityZone] && subnetMatchesAz(subnet) {
-			supported, err := typeCache.azSupportsInstanceType(ctx, m.client, *subnet.AvailabilityZone, ec2Settings.InstanceType)
-			if err != nil {
-				return nil, errors.Wrapf(err, "can't get supported instance types for AZ '%s'", *subnet.AvailabilityZone)
-			}
-			if supported {
-				overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{SubnetId: subnet.SubnetId})
-			}
-			AZSet[*subnet.AvailabilityZone] = true
-		}
+	for _, subnet := range supportingSubnets {
+		overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{SubnetId: aws.String(subnet.SubnetID)})
```

Comment:
Forgive me if this does already happen somewhere: do we consider that not all subnets are in the same region? i.e. for Fleet we'd only want to include us-east-1 availability zones.

Reply:
that's a great point.
fixed